common4j/src/main/com/microsoft/identity/common/java/telemetry/relay/AbstractTelemetryRelayClient.java

```diff
@@ -0,0 +1,108 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.common.java.telemetry.relay;
+
+import com.microsoft.identity.common.java.logging.Logger;
+import com.microsoft.identity.common.java.telemetry.observers.ITelemetryObserver;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * A relay client gives the flexibility to send telemetry events to a database.
+ * It extends a {@link ITelemetryObserver} and applies filter on every event captured before relaying the event to the database.
+ *
+ * @param <T>
+ */
+public abstract class AbstractTelemetryRelayClient<T> implements ITelemetryObserver<T> {
+
+    private static final String TAG = AbstractTelemetryRelayClient.class.getSimpleName();
+    // Filters to apply on the events captured.
+    // This basically performs an AND operation on the filters to determine whether the event will be relayed.
+    private final List<ITelemetryEventFilter<T>> eventFilters = new ArrayList<>();
+
+    @Override
+    public void onReceived(T telemetryData) {
+        final String methodTag = TAG + ":onReceived";
+
+        for (final ITelemetryEventFilter<T> filter : eventFilters) {
+            // Prevent event relay if any of the filters returns false.
+            if (!filter.shouldRelay(telemetryData)) {
+                return;
+            }
+        }
+
+        try {
+            relayEvent(telemetryData);
+        } catch (TelemetryRelayException e) {
+            Logger.error(methodTag, "Error relaying telemetry data", e);
+        }
+    }
+
+    /**
+     * Remove an event filter from the list of filters
+     */
+    public void removeFilter(ITelemetryEventFilter<T> filter) {
+        this.eventFilters.remove(filter);
+    }
+
+    /**
+     * Add an event filter to the list of filters
+     */
+    public void addFilter(ITelemetryEventFilter<T> filter) {
+        this.eventFilters.add(filter);
+    }
+
+    /**
+     * Clear all the filters in this relay client
+     */
+    public void clearFilters() {
+        this.eventFilters.clear();
+    }
+
+    /**
+     * Initialize the relay client. This should only happen once.
+     */
+    public abstract void initialize() throws TelemetryRelayException;
+
+    /**
+     * Returns true if the relay client has been initialized.
+     */
+    public abstract boolean isInitialized();
+
+    /**
+     * Invoked when an event is ready to be relayed
+     */
+    public abstract void relayEvent(final T eventData) throws TelemetryRelayException;
+
+    /**
+     * Flush any pending telemetry events in memory to disk and shutdown the telemetry system.
+     * This method can be invoked when the application is being closed.
+     */
+    public abstract void flushAndTeardown();
+
+    /**
+     * Flush any pending telemetry events in memory to disk.
+     */
+    public abstract void flush();
```

Comment:
Which of these are we using currently in our code?

Reply:
Is this the same `flush` in the CommandDispatcher flow? or are we calling somewhere else too?

Reply:
https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/1757/commits/49bfef79d16c8e9616933edb33160b6bfe0695e5